swupd/delta.go

```diff
@@ -0,0 +1,160 @@
+	"github.com/pkg/errors"
+)
+
+// CreateDeltas creates all delta files between the previous and current
+// version of the supplied manifest, returning a list of files that
+// delta creation failed for.
+func CreateDeltas(manifest *Manifest, statedir string) ([]*File, error) {
+	var c config
+	c, err := getConfig(statedir)
+	if err != nil {
+		return nil, err
+	}
+
+	var oldManifest *Manifest
+	var newManifest *Manifest
+
+	if oldManifest, err = ParseManifestFile(fmt.Sprintf("%s/%d/%s", c.outputDir, manifest.Header.Previous, manifest.Name)); err != nil {
+		return nil, err
+	}
+	if newManifest, err = ParseManifestFile(fmt.Sprintf("%s/%d/%s", c.outputDir, manifest.Header.Version, manifest.Name)); err != nil {
+		return nil, err
+	}
+
+	_, _, _ = newManifest.linkPeersAndChange(oldManifest, 0)
+
+	deltas, err := consolidateDeltaFiles(newManifest, c)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to create deltas list %s", manifest.Name)
+	}
+
+	var failed []*File
+	var errchan = make(chan *File)
+	var wg sync.WaitGroup
+	for _, file := range deltas {
+		wg.Add(1)
+		go func(file *File, statedir string) {
+			defer wg.Done()
+			err := createDelta(file, statedir, file.DeltaPeer.Version, file.DeltaPeer.Hash)
+			if err != nil {
+				errchan <- file
+			}
+		}(file, statedir)
+	}
+
+	go func() {
+		wg.Wait()
+		close(errchan)
+	}()
+
+	for f := range errchan {
+		failed = append(failed, f)
+	}
+
+	return failed, nil
+}
+
+func consolidateDeltaFiles(manifest *Manifest, c config) ([]*File, error) {
+	var files []*File
+
+	if manifest == nil {
+		return nil, nil
+	}
+
+	for _, file := range manifest.Files {
+		if file.Version <= manifest.Header.Version ||
+			file.DeltaPeer == nil ||
+			(file.Type != typeFile && file.Type != typeDirectory && file.Type != typeLink) {
+			continue
+		}
+
+		deltaFile := fmt.Sprintf("%s/%d/delta/%d-%d-%s-%s",
+			c.outputDir, file.Version, file.DeltaPeer.Version,
+			file.Version, file.DeltaPeer.Hash, file.Hash)
+
+		// Only add to list if the delta file does not already exist on the
+		// system or in the list
+		if _, err := os.Stat(deltaFile); os.IsNotExist(err) {
+			if found := file.findFileNameInSlice(files); found == nil {
+				// Prepend to list...
+				files = append([]*File{file}, files...)
+			}
+		}
+
+	}
+
+	return files, nil
+}
+
+func createDelta(file *File, statedir string, fromVersion uint32, fromHash Hashval) error {
+	var c config
+	c, err := getConfig(statedir)
+	if err != nil {
+		return err
+	}
+
+	// File without peers cannot have deltas
+	if file.DeltaPeer == nil {
+		return nil
+	}
+
+	// We only support deltas between two regular files for now
+	if file.Type != typeFile || file.DeltaPeer.Type != typeFile {
+		return nil
+	}
+
+	// By definition we cannot make deltas for files that are deleted
+	if file.Status == statusDeleted {
+		return nil
+	}
+
+	// TODO: Feature to support in the future, this is a complex case
+	if file.Type == typeDirectory {
```

Comment:
Except you already return if `file.Type != typeFile` above. Maybe move the TODO up there?